File: Kotlin/Ktor/src/main/kotlin/com/degenhf/auth/Application.kt
package com.degenhf.auth

import io.ktor.http.*
import io.ktor.serialization.jackson.*
import io.ktor.server.application.*
import io.ktor.server.auth.*
import io.ktor.server.auth.jwt.*
import io.ktor.server.engine.*
import io.ktor.server.netty.*
import io.ktor.server.plugins.contentnegotiation.*
import io.ktor.server.plugins.cors.routing.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import org.slf4j.LoggerFactory
import java.time.Duration

/**
 * Main Ktor application for ECC-based authentication
 */
fun main() {
    embeddedServer(Netty, port = 8080, host = "0.0.0.0") {
        module()
    }.start(wait = true)
}

fun Application.module() {
    val logger = LoggerFactory.getLogger("Application")

    // Install content negotiation
    install(ContentNegotiation) {
        jackson()
    }

    // Install CORS
    install(CORS) {
        allowMethod(HttpMethod.Options)
        allowMethod(HttpMethod.Get)
        allowMethod(HttpMethod.Post)
        allowMethod(HttpMethod.Put)
        allowMethod(HttpMethod.Delete)
        allowHeader(HttpHeaders.Authorization)
        allowHeader(HttpHeaders.ContentType)
        anyHost()
    }

    // Configure JWT authentication
    install(Authentication) {
        jwt("auth-jwt") {
            verifier(JwtConfig.verifier)
            validate { credential ->
                if (credential.payload.getClaim("username").asString() != null) {
                    JWTPrincipal(credential.payload)
                } else {
                    null
                }
            }
        }
    }

    // Initialize ECC auth handler
    val authOptions = EccAuthOptions(
        hashIterations = 100000,
        tokenExpiry = Duration.ofHours(24),
        cacheSize = 10000,
        cacheTtl = Duration.ofMinutes(5)
    )

    val authHandler = EccAuthHandler(authOptions)

    // Configure routing
    routing {
        // Health check
        get("/health") {
            call.respond(mapOf("status" to "healthy", "service" to "degenhf-ktor"))
        }

        // Public auth routes
        route("/api/auth") {
            // Register endpoint
            post("/register") {
                try {
                    val request = call.receive<RegisterRequest>()

                    if (request.username.isBlank() || request.password.length < 8) {
                        call.respond(HttpStatusCode.BadRequest, mapOf(
                            "error" to "Invalid input",
                            "message" to "Username must not be blank and password must be at least 8 characters"
                        ))
                        return@post
                    }

                    val userId = authHandler.register(request.username, request.password)
                    call.respond(HttpStatusCode.Created, mapOf(
                        "user_id" to userId,
                        "message" to "User registered successfully"
                    ))

                    logger.info("User registered: ${request.username}")

                } catch (e: Exception) {
                    logger.error("Registration failed", e)
                    call.respond(HttpStatusCode.BadRequest, mapOf(
                        "error" to "Registration failed",
                        "message" to e.message
                    ))
                }
            }

            // Login endpoint
            post("/login") {
                try {
                    val request = call.receive<LoginRequest>()

                    if (request.username.isBlank() || request.password.isBlank()) {
                        call.respond(HttpStatusCode.BadRequest, mapOf(
                            "error" to "Invalid input",
                            "message" to "Username and password are required"
                        ))
                        return@post
                    }

                    val token = authHandler.authenticate(request.username, request.password)
                    call.respond(HttpStatusCode.OK, mapOf(
                        "token" to token,
                        "message" to "Login successful"
                    ))

                    logger.info("User logged in: ${request.username}")

                } catch (e: Exception) {
                    logger.error("Login failed", e)
                    call.respond(HttpStatusCode.Unauthorized, mapOf(
                        "error" to "Authentication failed",
                        "message" to e.message
                    ))
                }
            }
        }

        // Protected routes
        authenticate("auth-jwt") {
            route("/api/auth") {
                // Verify token endpoint
                get("/verify") {
                    try {
                        val principal = call.principal<JWTPrincipal>()!!
                        val username = principal.payload.getClaim("username").asString()

                        call.respond(HttpStatusCode.OK, mapOf(
                            "user_id" to principal.payload.subject,
                            "username" to username,
                            "message" to "Token is valid"
                        ))

                        logger.debug("Token verified for user: $username")

                    } catch (e: Exception) {
                        logger.error("Token verification failed", e)
                        call.respond(HttpStatusCode.Unauthorized, mapOf(
                            "error" to "Token verification failed",
                            "message" to e.message
                        ))
                    }
                }

                // Get user profile
                get("/profile") {
                    try {
                        val principal = call.principal<JWTPrincipal>()!!
                        val userId = principal.payload.subject

                        val profile = authHandler.getUserProfile(userId)
                        call.respond(HttpStatusCode.OK, mapOf(
                            "user_id" to profile.userId,
                            "username" to profile.username,
                            "profile" to mapOf(
                                "created_at" to profile.createdAt.toString(),
                                "last_login" to profile.lastLogin.toString()
                            )
                        ))

                        logger.debug("Profile retrieved for user: ${profile.username}")

                    } catch (e: Exception) {
                        logger.error("Profile retrieval failed", e)
                        call.respond(HttpStatusCode.InternalServerError, mapOf(
                            "error" to "Profile retrieval failed",
                            "message" to e.message
                        ))
                    }
                }
            }

            // Example protected route
            get("/api/protected") {
                val principal = call.principal<JWTPrincipal>()!!
                val username = principal.payload.getClaim("username").asString()

                call.respond(HttpStatusCode.OK, mapOf(
                    "message" to "Welcome to protected route, $username!",
                    "user_id" to principal.payload.subject,
                    "timestamp" to System.currentTimeMillis()
                ))

                logger.debug("Protected route accessed by: $username")
            }
        }
    }

    logger.info("DegenHF Ktor server started on port 8080")
}

/**
 * JWT Configuration
 */
object JwtConfig {
    private const val secret = "degenhf-ktor-secret-key-change-in-production"
    private const val issuer = "degenhf-ktor"
    private const val audience = "degenhf-users"

    val verifier = io.jsonwebtoken.JwtParserBuilder()
        .setSigningKey(io.jsonwebtoken.security.Keys.hmacShaKeyFor(secret.toByteArray()))
        .setIssuer(issuer)
        .setAudience(audience)
        .build()
}

/**
 * Request/Response data classes
 */
data class RegisterRequest(
    val username: String,
    val password: String
)

data class LoginRequest(
    val username: String,
    val password: String
)

File: Languages/Swift/Vapor/Sources/App/Application.swift
import Vapor
import JWT

// MARK: - Main Application

@main
struct App {
    static func main() async throws {
        let app = try await Application.make()

        // Configure application
        try configure(app)

        // Run application
        try await app.execute()
    }
}

// MARK: - Configuration

func configure(_ app: Application) async throws {
    // Logging
    app.logger.logLevel = .info

    // JWT Configuration
    await configureJWT(app)

    // Middleware
    app.middleware.use(CORSMiddleware(configuration: .default()))
    app.middleware.use(app.sessions.middleware)

    // Routes
    try routes(app)

    // Initialize auth handler
    let authOptions = EccAuthOptions(
        hashIterations: 100000,
        tokenExpiry: 86400, // 24 hours
        cacheSize: 10000,
        cacheTtl: 300 // 5 minutes
    )

    let authHandler = EccAuthHandler(options: authOptions)
    app.storage.set(EccAuthHandlerKey.self, to: authHandler)

    app.logger.info("DegenHF Vapor server configured and ready")
}

// MARK: - JWT Configuration

func configureJWT(_ app: Application) async {
    // In production, load from secure key management
    let privateKey = P256.Signing.PrivateKey()

    await JWTSignersStorage.configure(app, signers: [
        .es256(key: privateKey)
    ], kid: "degenhf-vapor")
}

// MARK: - Routes

func routes(_ app: Application) throws {
    let authHandler = app.storage.get(EccAuthHandlerKey.self)!

    // Health check
    app.get("health") { req async -> Response in
        let response = Response(status: .ok)
        try response.content.encode([
            "status": "healthy",
            "service": "degenhf-vapor",
            "timestamp": Int(Date().timeIntervalSince1970)
        ])
        return response
    }

    // Public auth routes
    let authRoutes = app.grouped("api", "auth")

    // Register endpoint
    authRoutes.post("register") { req async throws -> Response in
        let registerRequest = try req.content.decode(RegisterRequest.self)

        guard !registerRequest.username.isEmpty else {
            throw Abort(.badRequest, reason: "Username cannot be empty")
        }
        guard registerRequest.password.count >= 8 else {
            throw Abort(.badRequest, reason: "Password must be at least 8 characters")
        }

        do {
            let userId = try authHandler.register(username: registerRequest.username, password: registerRequest.password)
            let response = Response(status: .created)
            try response.content.encode([
                "user_id": userId,
                "message": "User registered successfully"
            ])
            req.logger.info("User registered: \(registerRequest.username)")
            return response
        } catch let error as AuthError {
            req.logger.error("Registration failed: \(error)")
            throw Abort(.badRequest, reason: error.localizedDescription)
        }
    }

    // Login endpoint
    authRoutes.post("login") { req async throws -> Response in
        let loginRequest = try req.content.decode(LoginRequest.self)

        guard !loginRequest.username.isEmpty else {
            throw Abort(.badRequest, reason: "Username cannot be empty")
        }
        guard !loginRequest.password.isEmpty else {
            throw Abort(.badRequest, reason: "Password cannot be empty")
        }

        do {
            let token = try authHandler.authenticate(username: loginRequest.username, password: loginRequest.password)
            let response = Response(status: .ok)
            try response.content.encode([
                "token": token,
                "message": "Login successful"
            ])
            req.logger.info("User logged in: \(loginRequest.username)")
            return response
        } catch let error as AuthError {
            req.logger.error("Login failed: \(error)")
            throw Abort(.unauthorized, reason: error.localizedDescription)
        }
    }

    // Protected routes (require JWT authentication)
    let protectedRoutes = authRoutes.grouped(JWTMiddleware())

    // Verify token endpoint
    protectedRoutes.get("verify") { req async throws -> Response in
        let payload = try req.jwt.verify(as: UserPayload.self)

        let response = Response(status: .ok)
        try response.content.encode([
            "user_id": payload.subject.value,
            "username": payload.username,
            "message": "Token is valid"
        ])
        req.logger.debug("Token verified for user: \(payload.username)")
        return response
    }

    // Get user profile
    protectedRoutes.get("profile") { req async throws -> Response in
        let payload = try req.jwt.verify(as: UserPayload.self)
        let authHandler = req.application.storage.get(EccAuthHandlerKey.self)!

        do {
            let profile = try authHandler.getUserProfile(userId: payload.subject.value)
            let response = Response(status: .ok)
            try response.content.encode([
                "user_id": profile.userId,
                "username": profile.username,
                "profile": [
                    "created_at": ISO8601DateFormatter().string(from: profile.createdAt),
                    "last_login": ISO8601DateFormatter().string(from: profile.lastLogin)
                ]
            ])
            req.logger.debug("Profile retrieved for user: \(profile.username)")
            return response
        } catch let error as AuthError {
            req.logger.error("Profile retrieval failed: \(error)")
            throw Abort(.internalServerError, reason: error.localizedDescription)
        }
    }

    // Example protected route
    app.get("api", "protected") { req async throws -> Response in
        let payload = try req.jwt.verify(as: UserPayload.self)

        let response = Response(status: .ok)
        try response.content.encode([
            "message": "Welcome to protected route, \(payload.username)!",
            "user_id": payload.subject.value,
            "timestamp": Int(Date().timeIntervalSince1970)
        ])
        req.logger.debug("Protected route accessed by: \(payload.username)")
        return response
    }
}

// MARK: - JWT Middleware

struct JWTMiddleware: AsyncMiddleware {
    func respond(to request: Request, chainingTo next: AsyncResponder) async throws -> Response {
        guard let token = request.headers.bearerAuthorization?.token else {
            throw Abort(.unauthorized, reason: "Missing or invalid authorization header")
        }

        do {
            let payload = try request.jwt.verify(token, as: UserPayload.self)
            request.storage.set(UserPayload.self, to: payload)
            return try await next.respond(to: request)
        } catch {
            request.logger.error("JWT verification failed: \(error)")
            throw Abort(.unauthorized, reason: "Invalid token")
        }
    }
}

// MARK: - Storage Keys

struct EccAuthHandlerKey: StorageKey {
    typealias Value = EccAuthHandler
}

// MARK: - Request/Response Models

struct RegisterRequest: Content {
    let username: String
    let password: String
}

struct LoginRequest: Content {
    let username: String
    let password: String
}

// MARK: - JWT Payload Extension

extension Request {
    var jwt: JWTSigners {
        application.jwt.signers
    }
}